### PR TITLE
implemented preinit of jdbc connection pool

### PR DESCRIPTION
**Pull request by frschwab**
## Description

Hello JMeter developers,

I would like to suggest an enhancement to the JDBC connection configuration:
An option (true/false) where you can choose to pre-initialize the JDBC connection pool.
The default option should be "false", in order to not break existing behavior.

## Motivation and Context

fixes https://bz.apache.org/bugzilla/show_bug.cgi?id=63406

Currently, you get longer execution times for the first JDBC request for each thread of a thread group,
because the JDBC pool is only initialized when the first JDBC request asks the JDBC connection pool for a connection to execute its sql query.

Initialization currently also means: initializing the whole pool - before any JDBC request sampler is given a connection.
That means that each JDBC request, that is started before the connection pool is fully initialized, is affected by measuring connection establishment times for each of the following KPIs:
Elapsed time, latency, connect time.

A discussion took also here place already:
https://github.com/frschwab/jmeter/commit/102f0ac58af82a28e01565047f3d114366996231

## How Has This Been Tested?
not yet, please suggest how to test.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.
- [x] implement tests

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

**Comment by frschwab** [comment]
I just saw that I programmed the default value to be TRUE, instead of my description for the PR stating it should be FALSE.
So let us discuss this, I am open to any recommendations ;-)

**Comment by frschwab** [comment]
preinit_pool

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java [hunk 1]
```diff
@@ -289,6 +292,16 @@ private BasicDataSource initPool(String maxPool) {
...
 
         log.debug("PoolConfiguration:{}", this.dataSource);
+
+        if(isPreinit()) {
+            // side effect - connection pool init - that is what we want
+            try {
+                dataSource.getConnection().close();
+            } catch (SQLException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
```
Comment (pmouawad) [inline src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: e.printStackTrace() must be avoided and replaced by logging or in this case make it silent ?

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementBeanInfo.java [hunk 2]
```diff
@@ -82,6 +82,9 @@ public DataSourceElementBeanInfo() {
         Set<String> modesSet = TRANSACTION_ISOLATION_MAP.keySet();
         String[] modes = modesSet.toArray(new String[modesSet.size()]);
         p.setValue(TAGS, modes);
+        p = property("preinit");
+        p.setValue(NOT_UNDEFINED, Boolean.TRUE);
+        p.setValue(DEFAULT, Boolean.TRUE);
```
Comment (pmouawad) [inline src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementBeanInfo.java]: default to FALSE for backward compatibility ?

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementResources.properties [hunk 3]
```diff
@@ -46,3 +46,5 @@ keepAlive.displayName=Test While Idle
 keepAlive.shortDescription=Whether the pool should validate connections while they are idle in Pool.  If no, Soft Min Evictable Idle Time and Validation Query are ignored.
 transactionIsolation.displayName=Transaction Isolation
 transactionIsolation.shortDescription=Transaction Isolation Level
+preinit.displayName=Preinit Pool
```
Comment (pmouawad) [inline src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementResources.properties]: Provide translation in French

**Review by pmouawad — CHANGES_REQUESTED** [review]
Hello,
Thanks again for contributing.
Find some review notes.

Regards
Philippe

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 1 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: thanks, I changed it

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementBeanInfo.java (hunk 2 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementBeanInfo.java]: thanks, I changed it

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementResources.properties (hunk 3 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElementResources.properties]: thanks, I changed it

**Comment by pmouawad** [comment]
@frschwab  Thanks for taking into account my notes. Do you think you could provide a JUnit or shall I merge it like this ? 
Thanks

**Comment by frschwab** [comment]
Let me think about how such a JUnit test could look like.
Also, I didn't update the documentation for this until now, should I do that?

**Comment by ubikloadpack** [comment]
> Let me think about how such a JUnit test could look like.
> Also, I didn't update the documentation for this until now, should I do that?

it would be great

**Comment by frschwab** [comment]
I updated the documentation text. How is the screenshot from the JDBC connection configuration in the documentation updated? Is this done automatically, or is it a manual process?

**Comment by pmouawad** [comment]
> I updated the documentation text. How is the screenshot from the JDBC connection configuration in the documentation updated? Is this done automatically, or is it a manual process?
Hello, you put the image in a folder of xdocs/screenshots and reference it in your xml using figure tag.

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java [hunk 4]
```diff
@@ -288,6 +291,16 @@ private BasicDataSource initPool(String maxPool) {
             dataSource.setPassword(getPassword());
         }
 
+        if(isPreinit()) {
+            // side effect - connection pool init - that is what we want
+            try {
+                dataSource.getConnection().close();
+                log.debug("Preinitializing the connection pool: {}@{}", getDataSourceName(), System.identityHashCode(dataSource));
```
Comment (pmouawad) [inline src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: can you surround it with log.isDebugEnabled() ?

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java [hunk 5]
```diff
@@ -288,6 +291,16 @@ private BasicDataSource initPool(String maxPool) {
...
+            } catch (SQLException ex) {
+                log.error("Error preinitializing the connection pool: {}@{}", getDataSourceName(), System.identityHashCode(dataSource), ex);
```
Comment (pmouawad) [inline src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: same thing with log.isErrorEnabled()

**Review by pmouawad — CHANGES_REQUESTED** [review]

**Comment by frschwab** [comment]
Open work from my side:
- [ ] implement test(s)
- [ ] update screenshot for user manual
- [ ] logging improvements as suggested

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 4 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: Hi, I found a lot of examples where calls to log.debug() are not surrounded with log.isDebugEnabled(). Others are, though...  When to use which approach?

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 5 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: See my reply to log.isDebugEnabled().

**Comment by frschwab** [comment]
I would like to adapt https://github.com/apache/jmeter/blob/master/bin/testfiles/JDBC_TESTS.jmx and use it as a test for the preinitialization of the connection pool.
One test that e.g. creates a big amount of connections to that instance, and where the "select 1" statements  take quite long (when preinit=off) - compared to shorter response times when preinit=on.
So it is rather a "performance" test than a real unit test, as you would compare response times and decide if that is a failing or succeeding test.
What do you think?

**Comment by vlsi** [comment]
> So it is rather a "performance" test than a real unit test

I don't like that type of test.

Could the test be like "provide invalid DB connection parameters and verify that the test fails before running a single sampler"?

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 5 above)
Comment (vlsi) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: Errors should be rare, so I don't see a reason to use `isErrorEnabled` ever.

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 4 above)
Comment (vlsi) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: Technically speaking, `isDebugEnabled` should be used when "arguments have computations" or "varargs are used for logging".

In this case, `System.identityHashCode` will be computed no matter if `debug` is enabled or not, so it it makes sense to use `isDebugEnabled.

Even though this "connection pool initialization" will be used just once, the code might evolve (or even be copy-pasted"), so I think it is worth adding `isDebugEnabled` when the first statement holds.

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 5 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: Thanks, I also surrounded it with a check - others did it, too, in the source code, I found several examples.

File: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java (hunk 4 above)
Comment (frschwab) [reply src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java]: thanks, I surrounded it with a check

**Comment by frschwab** [comment]
> > So it is rather a "performance" test than a real unit test
> 
> I don't like that type of test.
> 
> Could the test be like "provide invalid DB connection parameters and verify that the test fails before running a single sampler"?

I used a groovy jdbc call into the test database to check if there is a preinitialized JDBC connection from the pool and added it to the JDBC_TESTS.jmx file.

From my side, I am done. Please merge! Thank you!

**Pull request merged**